File: apps/admin_app/lib/admin_app_web/live/components/vega_lite.ex
defmodule AdminAppWeb.VegaLiteComponent do
  use AdminAppWeb, :live_component

  @impl true
  def update(assigns, socket) do
    socket = assign(socket, id: assigns.id)
    # Send the specification object to the hook, where it gets
    # rendered using the client side Vega-Lite package
    {:ok, push_event(socket, "vega_lite:#{socket.assigns.id}:init", %{"spec" => assigns.spec})}
  end

  @impl true
  def render(assigns) do
    ~H"""
    <div id={"vega-lite-#{@id}"} phx-hook="VegaLite" phx-update="ignore" data-id={@id} >
    </div>
    """
  end
end

File: apps/admin_app/assets/js/hooks/vega_lite.js
import vegaEmbed from "vega-embed";

/**
 * A hook used to render graphics according to the given
 * Vega-Lite specification.
 *
 * The hook expects a `vega_lite:<id>:init` event with `{ spec }` payload,
 * where `spec` is the graphic definition as an object.
 *
 * Configuration:
 *
 *   * `data-id` - plot id
 */
const VegaLite = {
  mounted() {
    this.id = this.el.getAttribute("data-id");
    this.viewPromise = null;

    const container = document.createElement("div");
    this.el.appendChild(container);
    console.log(this)

    this.handleEvent(`vega_lite:${this.id}:init`, ({ spec }) => {
      console.log(spec)
      this.viewPromise = vegaEmbed(container, spec, {})
        .then((result) => result.view)
        .catch((error) => {
          console.error(
            `Failed to render the given Vega-Lite specification, got the following error:\n\n    ${error.message}\n\nMake sure to check for typos.`
          );
        });
    });
  },

  destroyed() {
    if (this.viewPromise) {
      this.viewPromise.then((view) => view.finalize());
    }
  },
};

export default VegaLite;
